### PR TITLE
Fix crash on missing DEF frames for in-game assets

### PR DESCRIPTION
**Pull request by MichalZr6**
This PR addresses the issue #4547 where the game crashes when attempting to load DEF frames for in-game assets, particularly when the specified frames are missing or invalid. While the underlying functionality of using DEF frames as pictures is currently disabled (due to changes to xbrz/upscaled support), this fix ensures the game handles missing frames without crashing. You will only see log message, for example:

> Frame index 50 is out of bounds for group 0 in file: SPRITES/TOWN-SCREEN/CASTLE/HALLCSTL

Please note: I have not tested this extensively, and the fix might not work in all cases.

### CODE REVIEW

**Comment by IvanSavenko** [comment]
I think way easier and simple logic is to check whether DefFile contains specified frame, and if not - return "DEFAULT" image in `RenderHandler::loadImageFromFileUncached`

**Comment by MichalZr6** [comment]
What do you mean by default image to return ? Something like this ?
```cpp
if(defFile->hasFrame(locator.defFrame, locator.defGroup))
	return std::make_shared<SDLImageShared>(defFile.get(), locator.defFrame, locator.defGroup, preScaledFactor);
else
{
	logGlobal->error("Frame %d in group %d not found in file: %s", 
		locator.defFrame, locator.defGroup, locator.defFile->getName().c_str());
	return std::make_shared<SDLEmptyImageShared>();
}
```

```cpp
class SDLEmptyImageShared : public SDLImageShared
{
public:
	SDLEmptyImageShared() : SDLImageShared(createDefaultSurface().get()) {}

private:
	std::shared_ptr<SDL_Surface> createDefaultSurface()
	{
		auto surface = SDL_CreateRGBSurface(0, 32, 32, 32, 0, 0, 0, 0);
		SDL_FillRect(surface, nullptr, SDL_MapRGB(surface->format, 255, 0, 0));
		return std::shared_ptr<SDL_Surface>(surface, SDL_FreeSurface);
	}
};
```
the size would have to be probably acquired from defFile with some additional code in this class for retrieving it, but not a problem.

**Comment by IvanSavenko** [comment]
What I mean is add something like this to `RenderHandler::loadImageFromFileUncached`:
```cpp
		if (defFile->containsImage(locator.defFrame, locator.defGroup))
			return std::make_shared<SDLImageShared>(defFile.get(), locator.defFrame, locator.defGroup, preScaledFactor);
		else
			return std::make_shared<SDLImageShared>(ImagePath::builtin("DEFAULT"), locator.preScaledFactor);
```
(`containsImage` method does not exists yet, and needs to be implemented)

Also - method `RenderHandler::getLocatorForAnimationFrame` already has check for non-existing frame in def. Why that check is not triggered in this scenario?

**Comment by MichalZr6** [comment]
Thanks. I have already implemented `containsImage` on my local branch (named it `hasFrame` instead) but didn't know what the default image should be. With `ImagePath::builtin("DEFAULT")` the town hall screen (with missing img) looks like this:

obraz

As for `getLocatorForAnimationFrame` - it never triggers. Both in my test case with missing frame and also while clicking on some creatures, dwellings, running some battles, etc. Looked through call stack. Another function is used (`CAnimation::getImageLocator`) in [CAnimation::loadFrame](https://github.com/vcmi/vcmi/blob/150f4904aba88b9d1a9896adc0a863fb2c8a82ac/client/render/CAnimation.cpp#L21)

File: client/renderSDL/SDLImage.h [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 #include "../render/IImage.h"
 #include "../../lib/Point.h"
+#include <SDL_surface.h>
```
Comment (IvanSavenko) [inline client/renderSDL/SDLImage.h:14]: Please avoid includes in headers, especially 3rd party libraries

File: client/renderSDL/SDLImage.h [hunk 2]
```diff
@@ -66,6 +67,20 @@ class SDLImageShared final : public ISharedImage, public std::enable_shared_from
...
+
+private:
+	std::shared_ptr<SDL_Surface> createDefaultSurface()
+	{
+		auto surface = SDL_CreateRGBSurface(0, 32, 32, 32, 0, 0, 0, 0);
+		SDL_FillRect(surface, nullptr, SDL_MapRGB(surface->format, 255, 0, 0));
+		return std::shared_ptr<SDL_Surface>(surface, SDL_FreeSurface);
+	}
+};
+
```
Comment (IvanSavenko) [inline client/renderSDL/SDLImage.h:83]: Unused? 
Also - what's exact point of this class?

If you're implementing my TODO, then its goal was somewhat different - it was to create dummy implementation of ISharedImage to replace checks for non-null surface in every method of SDLImageShared

File: client/render/CDefFile.cpp [hunk 3]
```diff
@@ -38,7 +38,8 @@ enum class DefType : uint32_t
 
 CDefFile::CDefFile(const AnimationPath & Name):
 	data(nullptr),
-	palette(nullptr)
+	palette(nullptr),
+	fileName(Name.getName())
```
Comment (IvanSavenko) [inline client/render/CDefFile.cpp:42]: Unused?

**Review by IvanSavenko — COMMENTED** [review]
I think all changes in SDLImage.h can be safely reverted?

File: client/renderSDL/SDLImage.h (hunk 2 above)
Comment (MichalZr6) [reply client/renderSDL/SDLImage.h]: Just tried to implement some "default" as I didn't realise that there is sth like that already.

**Comment by MichalZr6** [comment]
> I think all changes in SDLImage.h can be safely reverted?

Reverted unnecessary changes.

**Pull request merged**